compositor_render/src/transformations/builtin/box_layout.rs

```diff
@@ -10,6 +14,45 @@ pub struct BoxLayout {
 }
 
 impl BoxLayout {
+    /// Returns box representing position of input frame fitted into `self`,     
+    /// Fitted means scaled to input resolution without cropping or changing aspect ratio
+    pub fn fit(
+        &self,
+        input_resolution: Resolution,
+        x_align: HorizontalAlign,
+        y_align: VerticalAlign,
+    ) -> BoxLayout {
+        let input_x_scale = self.width / input_resolution.width as f32;
+        let input_y_scale = self.height / input_resolution.height as f32;
+
+        let input_scale = input_x_scale.min(input_y_scale);
+
+        let x_padding = self.width - input_resolution.width as f32 * input_scale;
+        let y_padding = self.height - input_resolution.height as f32 * input_scale;
+
+        let left_padding = match x_align {
+            HorizontalAlign::Left => 0.0,
+            HorizontalAlign::Right => x_padding,
+            HorizontalAlign::Center | HorizontalAlign::Justified => x_padding / 2.0,
+        };
+
+        let top_padding = match y_align {
+            VerticalAlign::Top => 0.0,
+            VerticalAlign::Center => y_padding / 2.0,
+            VerticalAlign::Bottom => y_padding,
+        };
+
+        BoxLayout {
+            top_left_corner: (
+                self.top_left_corner.0 + left_padding,
+                self.top_left_corner.1 + top_padding,
+            ),
+            width: self.width - x_padding,
+            height: self.height - y_padding,
+            rotation_degrees: self.rotation_degrees,
+        }
+    }
```

Comment:
This probably has less numerical precision, than the old approach, but since it's significantly simpler, self-documenting, and doesn't require any thinking at all, I decided to leave it this way. With floats, this shouldn't have any visible effect anyway.